k2/csrc/rnnt_decode.h

```diff
@@ -0,0 +1,390 @@
+#include "k2/csrc/array_ops.h"
+#include "k2/csrc/log.h"
+#include "k2/csrc/macros.h"
+
+namespace k2 {
+namespace rnnt_decoding {
+
+/*
+  The RNN-T decoding implemented here is for what we call "modified" RNN-T, or
+  equivalently, regular RNN-T but with max_sym_per_frame set to 1.  (We can
+  train with the "modified" option set with some probability, in order to ensure
+  that the trained model is compatible with decoding with this setting).
+
+  - contexts are finite symbol left-contexts, of length
+  RnntDecodingConfig::decoder_history_len. conceptuatly they represent a list of
+  `decoder_history_len` symbols; they are represented numerically as, for
+  example in the length-2-history case: symbol_{t-1} +  symbol_{t-2} *
+  vocab_size.
+
+  - frames come from the transcription network, they are derived from
+  sub-sampling of acoustic frames or samples.
+ */
+
+struct RnntDecodingConfig {
+  RnntDecodingConfig(int32_t vocab_size, int32_t decoder_history_len,
+                     double beam, int32_t max_states, int32_t max_contexts)
+      : vocab_size(vocab_size),
+        decoder_history_len(decoder_history_len),
+        beam(beam),
+        max_states(max_states),
+        max_contexts(max_contexts) {
+    num_context_states = pow(vocab_size, decoder_history_len);
+  }
+
+  // vocab_size is the largest-symbol plus one.
+  int32_t vocab_size;
+
+  // decoder_history_len is the number of symbols
+  // of history the decoder takes; will normally
+  // be one or two ("stateless decoder"), this
+  // RNN-T decoding setup does not support
+  // unlimited decoder context such as with LSTMs
+  int32_t decoder_history_len;
+
+  // num_context_states == pow(vocab_size, decoder_history_len).
+  // We need an unique id for each context state, think about that if vocab_size
+  // equals to 10, we need 0 ~ 9 to distinguish each context state when
+  // decoder_history_len is 1, and 0 ~ 99 (10 ^ 2 ids) for decoder_history_len
+  // equals to 2, 0 ~ 999 (10 ^ 3 ids) for decoder_history_len equals to 3.
+  int32_t num_context_states;
+
+  // `beam` imposes a limit on the score of a state, relative to the
+  // best-scoring state on the same frame.  E.g. 10.
+  double beam;
+
+  // `max_states` is a limit on the number of distinct states that we allow per
+  // frame, per stream; the number of states will not be allowed to exceed
+  // this limit.
+  int32_t max_states;
+
+  // `max_contexts` is a limit on the number of distinct contexts that we allow
+  // per frame, per stream; the number of contexts will not be allowed to
+  // exceed this limit.
+  int32_t max_contexts;
+};
+
+struct ArcInfo {
+  // The arc-index within the RnntDecodingStream::graph that corresponds to this
+  // arc, or -1 if this arc is a "termination symbol" (these do not appear in
+  // the graph).
+  int32_t graph_arc_idx01;
+
+  // The score on the arc; contains both the graph score (if any) and the score
+  // from the RNN-T joiner.
+  float score;
```

Comment:
Should its type be `double`? I find that `scores` of ragged arrays in this file have `double` elements.


By the way, it will not increase memory usage if you order them in the following way:
```cpp
double score;
int32_t graph_arc_idx01;
int32_t dest_state;
```